[ADD] 10.0 base_jsonify

PierrickBrun opened the pull request:
Add module base_jsonify to repository.

This module add the jsonify method to the ORM. This method take as argument
the browse record and the "parser" that specify the field to extract.

Example of parser:



    parser = [
        'name',
        'number',
        'create_date',
        ('partner_id', ['id', 'display_name', 'ref'])
        ('line_id', ['id', ('product_id', ['name']), 'price_unit'])
    ]

In order to be consitent with the odoo api the jsonify method always
return a list of object even if there is only one element in input

By default the key into the json is the name of the field extracted
from the model. If you need to specify an alternate name to use as key, you
can define your mapping as follow into the parser definition:


    parser = [
         'field_name:json_key'
    ]



    parser = [
        'name',
        'number',
        'create_date:creationDate',
        ('partner_id:partners', ['id', 'display_name', 'ref'])
        ('line_id:lines', ['id', ('product_id', ['name']), 'price_unit'])
    ]

Also the module provide a method "get_json_parser" on the ir.exports object
that generate a parser from an ir.exports configuration

sebastienbeau committed:
add base_jsonify module

sebastienbeau committed:
if '.id' is used replaced it by id

sebastienbeau committed:
fix getting value of many2one if the value is None

sebastienbeau committed:
fix complatibility with 8

sebastienbeau committed:
fix travis

lmignon committed:
base_jsonify: Allow fields aliasing in jsonify

lmignon committed:
base_jsonify: typo

sebastienbeau committed:
add support of reference fields

sebastienbeau committed:
remove False for int, float, string

hparfr committed:
Remove copyrights from init and update copyrights

hparfr committed:
Fix tests and add fix nullable fields

hparfr committed:
from openerp -> from odoo<

hparfr committed:
Fix monkey patchin

lmignon committed:
base_jsonify: Output 'false' into json only for boolean fields

sebastienbeau committed:
restore travis configuration (#112)

* [ADD] restore travis configuration

* [PEP] fix pep8

* [REF] exclude  product_categ_available_pos of testing due to issue https://github.com/odoo/odoo/pull/23749

* [PEP] fix pep8

sbidoul reviewed (commented):
Could you use the readme fragments mechanism?

Commit:
Use readme fragment mechanism

PierrickBrun commented:
Hi @sbidoul , The module uses the readme fragments mechanism now, can you re-check and approve the PR ?

lmignon reviewed (changes requested):
@PierrickBrun It seems that the proposed sources is not the last one available on the original repo...  A least the know issue has been resolved by https://github.com/akretion/ak-odoo-incubator/commit/7297f5418830e01cbc6e6a5adc4724801448e689#diff-06d91370052ae6bd8a15b06f6c613182

Moreover, IMO we must keep the commit history from the original repo. You can find different ways to keep the history when moving one folder from one repo to an other one.
I've experienced yesterday a way to do it based on `git format-patch`

1. into the original repo:
    ```
    git format-patch --keep-subject  --root 10.0 -- base_jsonify
    ```
    The command will generate a patch file for each commit involving base_jsonify. (0001-*.patch,0002-*.patch,0003-*.patch............)

2. Checkout the 10.0 branch of https://github.com/OCA/server-tools and create a new branch
3. Into server-tools, apply each patch generated into you orginal repository with
    ```
    git am ../ak-odoo-incubator/*.patch
    ```

PierrickBrun commented:
Thank you for your detailed answer @lmignon , I've updated the PR with the forgotten fix and kept the history.

I was experimenting the same thing right now, I've found a similar but slightly different solution from this StackOverflow response : https://stackoverflow.com/a/11426261/7042184 

This is how I updated the PR. I don't know if this way is more adapted than yours but I think that edge cases have been well tested, judging from the discussion about it.

lmignon commented on base_jsonify/readme/ROADMAP.rst:
```diff
@@ -0,0 +1 @@
+Make a distinction between float 0.0L and no value
```
@PierrickBrun IMO this issue has been solved by 3ca15ceada6c48ac9e6a87f41c41719dcfe4ca89

PierrickBrun replied:
indeed

lmignon reviewed (changes requested):
The roadmap is no more valid and should be removed...

legalsylvain committed:
IMP add view for ir.exports

simahawk commented on base_jsonify/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+# -*- coding: utf-8 -*-
+# © 2017 Akretion (http://www.akretion.com)
```
Copyright

simahawk commented on base_jsonify/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+    "summary": "Base module that provide the jsonify method on all object",
+    "version": "10.0.1.0.0",
+    "category": "Uncategorized",
+    "website": "https://odoo-community.org/",
```
https://github.com/OCA/server-tools

simahawk commented on base_jsonify/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+
+{
+    "name": "Base Jsonify",
+    "summary": "Base module that provide the jsonify method on all object",
```
on all models

simahawk commented on base_jsonify/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "external_dependencies": {
```
empty, remove it

simahawk commented on base_jsonify/models/ir_export.py:
```diff
@@ -0,0 +1,44 @@
+from odoo import api, models
+
+
+def update_dict(data, fields):
```
docstring

simahawk commented on base_jsonify/models/ir_export.py:
```diff
@@ -0,0 +1,44 @@
+        update_dict(data[field], fields[1:])
+
+
+def convert_dict(dict_parser):
```
docstring

simahawk commented on base_jsonify/models/ir_exports_line.py:
```diff
@@ -0,0 +1,35 @@
+    )
+
+    @api.constrains('alias', 'name')
+    def _check_alias(self):
```
docstring

simahawk commented on base_jsonify/models/ir_exports_line.py:
```diff
@@ -0,0 +1,35 @@
+
+    alias = fields.Char(
+        'Alias',
+        help='The complete path to the field where you can specify an '
```
please rephrase

simahawk replied:
it seems that `alias` is used only for the json parser conversion so why don't we call it `json_config` or `json_path` or something similar?

simahawk commented on base_jsonify/models/models.py:
```diff
@@ -0,0 +1,75 @@
+    _inherit = 'base'
+
+    @api.model
+    def __parse_field(self, parser_field):
```
any reason for double underscore?

simahawk replied:
docstring

simahawk commented on base_jsonify/models/models.py:
```diff
@@ -0,0 +1,75 @@
+
+    @api.multi
+    def jsonify(self, parser):
+        """ Convert the record according to the parser given
```
```"""Convert current recordset to JSON data using given parser.

[...] example [...]
put the example output right after
[...]
"""
```

simahawk commented on base_jsonify/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,42 @@
+This module add the jsonify method to the ORM. This method take as argument
```
```suggestion
This module adds the a `jsonify` method to every model in the ORM.
It works on the current recordset and requires a single argument `parser` that specifies the fields to extract.
```

simahawk commented on base_jsonify/models/models.py:
```diff
@@ -0,0 +1,75 @@
+                ('line_id', ['id', ('product_id', ['name']), 'price_unit'])
+            ]
+
+        In order to be consitent with the odoo api the jsonify method always
```
*consistent

Commit:
[FIX] misc. fixes in the comments/docstrings

simahawk reviewed (approved):
@sebastienbeau can we merge? :)

simahawk commented:
@lmignon @sebastienbeau @sbidoul piiiiiing :smile:

lmignon commented:
@simahawk merged! Thank you for the review
